Improve error reporting when Crossgen2 crashes on Managed C++

trylek opened the pull request:
Based on Manish's feedback from another PR I investigated options
to improve Crossgen2 diagnostics in the presence of Managed C++
that currently crashes the metadata reader in several places.
This is the barebones version of such support, centralizing detection
and reporting of the failures. Right now we don't have any pretty
way of reporting localized strings from this low-level code, I'll
welcome any suggestions how to polish the change in that direction.

Thanks

Tomas

/cc: @dotnet/crossgen-contrib

trylek committed:
Improve error reporting when Crossgen2 crashes on Managed C++

Based on Manish's feedback from another PR I investigated options
to improve Crossgen2 diagnostics in the presence of Managed C++
that currently crashes the metadata reader in several places.
This is the barebones version of such support, centralizing detection
and reporting of the failures. Right now we don't have any pretty
way of reporting localized strings from this low-level code, I'll
welcome any suggestions how to polish the change in that direction.

Thanks

Tomas

jkotas commented:
Should the managed C++ check rather be done up-front when we are loading the file?

trylek commented:
Hmm, at the very least that would make it easier to display a meaningful error message using the existing resource string mechanism ;-). Good idea, thanks, will do.

trylek committed:
Make Managed C++ scanning upfront per JanK's PR feedback

nattress commented on src/coreclr/src/tools/aot/crossgen2/Program.cs:
```diff
@@ -347,6 +348,27 @@ private int Run()
                         }
                     }
 
+                    foreach (string inputFilePath in allInputFilePaths.Values)
```
Would you mind breaking this out into a method? Ideally `Run` will get smaller over time

trylek replied:
Fixed in 3rd commit.

mangod9 commented on src/coreclr/src/tools/aot/crossgen2/Program.cs:
```diff
@@ -347,6 +348,27 @@ private int Run()
+                        EcmaModule module = typeSystemContext.GetModuleFromPath(inputFilePath);
+                        foreach (MetadataType type in module.GetAllTypes())
+                        {
+                            foreach (EcmaMethod method in type.GetAllMethods())
```
is this iteration over all methods lead to further startup concerns?

trylek replied:
Removed in 3rd commit.

nattress commented:
My 2 cents: I prefer the on-demand checking since inputs with managed C++ are probably the rare case but will pay the price for checking for it on startup always.

jkotas commented on src/coreclr/src/tools/aot/crossgen2/Program.cs:
```diff
@@ -347,6 +348,27 @@ private int Run()
+                    foreach (string inputFilePath in allInputFilePaths.Values)
+                    {
+                        EcmaModule module = typeSystemContext.GetModuleFromPath(inputFilePath);
+                        foreach (MetadataType type in module.GetAllTypes())
```
I did not meant to scan all methods and in all assemblies. That is very expensive. I meant to just check whether the assembly is managed C++ (you can do that by looking at PE header).

trylek replied:
So far I have only found the bit <code>CorFlags.NativeEntryPoint</code>. Is that a reliable Managed C++ indicator? Other than that I don't see any PE Characteristics, DllCharacteristics bit flags or assembly-level attributes usable for this detection. Are you aware of any existing code that performs this detection so that I can take a look and adapt it for use in CG2?

jkotas replied:
When both `COMIMAGE_FLAGS_ILONLY` and `COMIMAGE_FLAGS_IL_LIBRARY` bits in CorHeader are zero, it means the binary is MC++.

trylek replied:
Thanks Jan, fixed in 3rd commit.

trylek committed:
Fix Managed C++ check based on JanK's advice; Simon's feedback

I have changed the check to use the CorFlags.ILOnly / ILLibrary
combination (both bits must be zero for an assembly to be MC++).
Per Simon's suggestion I have moved the check outside of the
Run method. As part of that I have changed typeSystemContext
to a field on the Program class as a first step in breaking Run
into several smaller methods.

Thanks

Tomas
